### PR TITLE
Add Compilation_context.for_link_time_code_gen

### PR DESCRIPTION
**Pull request by rgrinberg**
This is tune with how we modify the compilation context for other
special schemes.

### CODE REVIEW

File: src/dune/link_time_code_gen.ml [hunk 1]
```diff
@@ -33,20 +33,8 @@ let generate_and_compile_module cctx ~precompiled_cmi ~name:basename ~lib ~code
        |> Option.value_exn |> Path.as_in_build_dir_exn
      in
      Build.write_file_dyn ml code);
-  let opaque =
-    Ocaml_version.supports_opaque_for_mli (Super_context.context sctx).version
```
Comment (rgrinberg) [inline src/dune/link_time_code_gen.ml]: @diml why do we compile with opaque (whenever it's available) here?

Comment (diml) [reply src/dune/link_time_code_gen.ml]: IIRC, it's because the mli was built with `-opaque` and so the implem needs to be built with `-opaque` as well.

Comment (rgrinberg) [reply src/dune/link_time_code_gen.ml]: Okay, thanks. I added a comment.

**Comment by rgrinberg** [comment]
@diml ping here

File: src/dune/compilation_context.mli [hunk 2]
```diff
@@ -75,3 +75,6 @@ val vimpl : t -> Vimpl.t option
 val modes : t -> Mode.Dict.Set.t
 
 val for_wrapped_compat : t -> t
+
+val for_link_time_code_gen :
```
Comment (diml) [inline src/dune/compilation_context.mli]:
```suggestion
val for_module_generated_at_link_time :
```

to make it clearer what the `module_` argument is about

**Pull request merged**